### PR TITLE
[Bugbash] fix speedup replacement related

### LINKED ISSUE #4905: aten::upsample_nearest2d is not Supported!
**Describe the issue**:
[2022-06-01 15:25:48] INFO (FixMaskConflict/MainThread) dim0 sparsity: 0.794792
[2022-06-01 15:25:48] INFO (FixMaskConflict/MainThread) dim1 sparsity: 0.000000
[2022-06-01 15:25:48] INFO (FixMaskConflict/MainThread) Dectected conv prune dim" 0
[2022-06-01 15:25:49] INFO (nni.compression.pytorch.speedup.compressor/MainThread) infer module masks...
[2022-06-01 15:25:49] INFO (nni.compression.pytorch.speedup.compressor/MainThread) Update mask for .aten::upsample_nearest2d.126
[2022-06-01 15:25:49] ERROR (nni.compression.pytorch.speedup.jit_translate/MainThread) aten::upsample_nearest2d is not Supported! Please report an issue at https://github.com/microsoft/nni. Thanks~
[2022-06-01 15:25:49] INFO (nni.compression.pytorch.speedup.compressor/MainThread) Update mask for fc
**Environment**:ubuntu18.04
- NNI version:2.7
- Training service (local|remote|pai|aml|etc):
- Client OS:
- Server OS (for remote mode only):
- Python version:3.9.7
- PyTorch/TensorFlow version:pytorch1.7.1
- Is conda/virtualenv/venv used?:conda
- Is running in Docker?:no

**Configuration**:
 - Experiment config (remember to remove secrets!):
 - Search space:

**Log message**:
 - nnimanager.log:
 - dispatcher.log:
 - nnictl stdout and stderr:
 ---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [11], in <cell line: 6>()
      4 # speedup the model
      5 from nni.compression.pytorch.speedup import ModelSpeedup
----> 6 ModelSpeedup(m, torch.rand(1, 3, 256, 256).to('cuda:0'), masks).speedup_model()

File /mnt/fu03/zhoulei/AN/lib/python3.9/site-packages/nni/compression/pytorch/speedup/compressor.py:512, in ModelSpeedup.speedup_model(self)
    509 fix_mask_conflict(self.masks, self.bound_model, self.dummy_input)
    511 _logger.info("infer module masks...")
--> 512 self.infer_modules_masks()
    513 _logger.info('resolve the mask conflict')
    515 # load the original stat dict before replace the model

File /mnt/fu03/zhoulei/AN/lib/python3.9/site-packages/nni/compression/pytorch/speedup/compressor.py:355, in ModelSpeedup.infer_modules_masks(self)
    353 curnode = visit_queue.get()
    354 # forward mask inference for curnode
--> 355 self.update_direct_sparsity(curnode)
    356 successors = self.torch_graph.find_successors(curnode.unique_name)
    357 for successor in successors:

File /mnt/fu03/zhoulei/AN/lib/python3.9/site-packages/nni/compression/pytorch/speedup/compressor.py:223, in ModelSpeedup.update_direct_sparsity(self, node)
    221         weight_mask = self.masks[module_name]
    222     _, module = get_module_by_name(self.bound_model, module_name)
--> 223     _auto_infer = AutoMaskInference(
    224         module, dummy_input, in_masks, weight_mask, in_constants=in_constants,
    225         state_dict=copy.deepcopy(module.state_dict()), batch_dim=self.batch_dim)
    226 self.auto_inferences[unique_name] = _auto_infer
    227 _auto_infer.name = node.unique_name

File /mnt/fu03/zhoulei/AN/lib/python3.9/site-packages/nni/compression/pytorch/speedup/infer_mask.py:80, in AutoMaskInference.__init__(self, module, dummy_input, in_masks, weight_mask, output_mask, name, in_constants, state_dict, batch_dim)
     76         self.in_masks[in_id] = torch.ones_like(self.dummy_input[in_id])
     77         # ones_like will put the created mask on the same device with the dummy_input
     78 
     79 # Initialize the mask for output tensors
---> 80 self.output = self.module(*dummy_input)
     81 # self.output.requires_grad_()
     82 if output_mask is not None:
     83     # assume the given output mask is right

File /mnt/fu03/zhoulei/AN/lib/python3.9/site-packages/torch/nn/modules/module.py:727, in Module._call_impl(self, *input, **kwargs)
    725     result = self._slow_forward(*input, **kwargs)
    726 else:
--> 727     result = self.forward(*input, **kwargs)
    728 for hook in itertools.chain(
    729         _global_forward_hooks.values(),
    730         self._forward_hooks.values()):
    731     hook_result = hook(self, input, result)

TypeError: forward() missing 1 required positional argument: 'input'

**How to reproduce it?**:

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
- fix a `customized_replace_func` bug, add information about expected replacement function contributions
- `PixelShuffle` can not be easily replaced, need more speedup function support. No replace for most situation for now.
- Cherry pick a fix from #4886
- Fix #4905 

### FollowUp ###
- need review discussion about all mapping-like op [view, reshape, ...]

### CODE REVIEW

**Review by aminabedi — APPROVED** [review]

File: nni/compression/pytorch/speedup/compressor.py [hunk 1]
```diff
@@ -459,13 +459,17 @@ def forward(self, x):
...
+                err_msg += f"you could implement module replacement by passing in"
+                err_msg += f"`customized_replace_func` to `{self.__class__.__name__}`. "
+                err_msg += f"You are welcome to contribute back to nni as native support if you have implemented the replacement function, "
+                err_msg += f"so that more users can benefit from your contributions."
+                raise RuntimeError(err_msg)
             _logger.info("replace module (name: %s, op_type: %s)",
                          g_node.name, m_type)
-            replace_function = replace_module[m_type]
-            if m_type in self.customized_replace_func:
-                replace_function = self.customized_replace_func[m_type]
```
Comment (zheng-ningxin) [inline nni/compression/pytorch/speedup/compressor.py:468]: Looks like exact same with `472` line? why writing in this way leads to a bug?

Comment (J-shang) [reply nni/compression/pytorch/speedup/compressor.py:468]: `m_type` may not in `replace_module`, will raise a `KeyError`

File: nni/compression/pytorch/speedup/compress_modules.py [hunk 2]
```diff
@@ -595,37 +596,37 @@ def replace_layernorm(layernorm, masks):
...
 
-    Returns
-    -------
-    torch.nn.PixelShuffle
-        The new pixelshuffle module
+    We can not replace pixelshuffle easily right now, pixelshuffle is a kind of location mapping.
+    It will map tensor with shape (r^2 * C, H, W) to (C, r * H, r* W). So we have a dependency here,
+    the preserved input channel number should be a multiple of C, and the multiple can be squared to positive integer.
+    This dependence is similar to the group dependency in ConvXD, but more restrictive,
+    i.e., each `r^2 input channels` group can not be free to preserve any number of channels, must be a number in [1, 4, 9, 16, ... , r^2].
```
Comment (QuanluZhang) [inline nni/compression/pytorch/speedup/compress_modules.py:605]: interesting, let's discuss this op

File: nni/compression/pytorch/speedup/jit_translate.py [hunk 3]
```diff
@@ -492,6 +492,36 @@ def forward(self, *args):
...
+
+        def forward(self, *args):
+            """
+            The first input of args is the target tensor to upsample
+            , the following parameters is useless, because we already
+            get the size_list and the scale_list by parsing the cpp_nodes.
+            """
+            return torch.nn.functional.upsample_nearest(args[0],
+                                                        size=self.size_list, scale_factor=self.scale_list)
+    print(node)
```
Comment (QuanluZhang) [inline nni/compression/pytorch/speedup/jit_translate.py:510]: what is this print used for?

File: nni/compression/pytorch/utils/shape_dependency.py [hunk 4]
```diff
@@ -85,6 +85,10 @@ def reshape_break_channel_dependency(op_node):
     """
     in_shape = op_node.auxiliary['in_shape']
     out_shape = op_node.auxiliary['out_shape']
+    if not in_shape or not out_shape:
+        return True
```
Comment (zheng-ningxin) [inline nni/compression/pytorch/utils/shape_dependency.py:89]: why in_shape/out_shape may be None? Should update the `graphutils.py` accordingly.

**Review by QuanluZhang — APPROVED** [review]

**Review by zheng-ningxin — APPROVED** [review]

**Pull request merged**